### PR TITLE
CUDA: Don't parse IR for modules with llvmlite

### PR DESCRIPTION
**Pull request by gmarkall**
(Note: this PR is based on #6725, as it triggers previously-unused paths that fail on all CUDA toolkits without it)

In preparation for using the bitcode interface to NVVM, this commit avoids parsing the IR for CUDA modules with llvmlite then dumping them back to a string - the rationale for this is that the compilation flow would have involved sending the IR through three different versions of LLVM:

1. LLVM 10 via llvmlite for parsing and dumping in the CodeLibrary
2. LLVM 7.0.0 to convert the LLVM 7 IR to LLVM 7 bitcode
3. LLVM 7.0.1 in NVVM

Removing the first step aims to reduce complexity and the potential for subtle issues, especially as llvmlite moves ahead to newer LLVM versions.

To support this change, the `CodeLibrary` is refactored slightly to split out a `CPUCodeLibrary` class whose methods still implement the first stage above, which is required for the CPU target.

A corresponding refactoring of the `BaseCPUCodegen` class is also made - the base class shared between CUDA and the CPU is `Codegen`. This cleans up the abstractions a little by eliding the need for the CUDA codegen to override some of the CPU codegen's methods to "hide" them.

This change mostly, but does not completely avoid parsing IR with LLVM 10 through llvmlite when using CUDA - some parsing is still required for `BaseContext.get_abi_sizeof`, which parses types to find out their size. This is expected to be both mostly harmless, and difficult to avoid, so it is left in place.

Some other remarks on the changes:

- The 32-bit CUDA triple is removed, since it is no longer supported.
- This commit also deprecates the `nvvm_options` kwarg for `inspect_ptx` as it seems to support inspecting PTX generated with options different to those that would normally be used when code is JIT-compiled, and it will hamper further planned reorganisation of the way code generation works in the CUDA target. It is not expected that it is widely used.
- The syntax for various stub functions (`___numba_atomic_double_add` etc.) is a little different coming from llvmlite compared to being dumped from LLVM 10, so the replacement strings are updated to account for this.
- llvmlite generates debuginfo flags in the old sttle (e.g. `isDefinition: true` instead of `DISPFlagDefinition`) so this substitution is no longer required.

### CODE REVIEW

**Comment by gmarkall** [comment]
@esc could this have a buildfarm run prior to review please? Given the nature of the changes I'd like to be able to check whether this needs some reworking before it goes through the whole review process... Many thanks in advance.

**Comment by stuartarchibald** [comment]
> @esc could this have a buildfarm run prior to review please? Given the nature of the changes I'd like to be able to check whether this needs some reworking before it goes through the whole review process... Many thanks in advance.

Buildfarm ID: `numba_smoketest_cuda_yaml_15`.

**Comment by stuartarchibald** [comment]
> > @esc could this have a buildfarm run prior to review please? Given the nature of the changes I'd like to be able to check whether this needs some reworking before it goes through the whole review process... Many thanks in advance.
> 
> Buildfarm ID: `numba_smoketest_cuda_yaml_15`.

This passed.

**Review by sklam — COMMENTED** [review]
Just some nitpicks and questions.

File: numba/cuda/compiler.py [hunk 1]
```diff
@@ -591,15 +591,18 @@ def __init__(self, py_func, argtypes, link=None, debug=False, inline=False,
             'opt': 3 if opt else 0
         }
 
-        llvm_ir = str(lib._final_module)
+        llvm_ir = [str(mod) for mod in lib.modules]
```
[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/cuda/compiler.py:594]: What would you rename it to? It's still LLVM IR.

File: numba/cuda/cudadrv/nvvm.py [hunk 2]
```diff
@@ -904,6 +892,8 @@ def set_cuda_kernel(lfunc):
     nmd = mod.get_or_insert_named_metadata('nvvm.annotations')
     nmd.add(md)
 
+    lfunc.linkage = 'external'
```
[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/cuda/cudadrv/nvvm.py]: I think this must be left over from when I was experimenting with compiling all modules in separate compilation units (in an attempt to fix some of the debug / lineinfo issues we have). I've tried removing it now and it seems to be OK for me too.

File: numba/core/codegen.py [hunk 3]
```diff
@@ -703,6 +761,7 @@ def finalize(self):
         self._final_module.verify()
         self._finalize_final_module()
 
+
```
[the remark this answers is not in the document]
Comment (gmarkall) [reply numba/core/codegen.py]: Removed.

**Comment by gmarkall** [comment]
@sklam Thanks for the review - made some changes accordingly (apart from renaming `llvm_ir`, because I'm not sure what you'd want it renaming to).

File: numba/cuda/compiler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sklam) [reply numba/cuda/compiler.py:594]: I was originally concerned about changing the type of `llvm_ir` and subsequently changing the interface in `CachedPTX`. But, now that I realize `CachedPTX` will be removed in https://github.com/numba/numba/pull/6769, I don't think the renaming will be necessary. It's just a temporary thing at this point.

**Review by sklam — APPROVED** [review]
Thanks for the PR. Looks good. I am sending it through the buildfarm again given the changes in the linkage. `numba_smoketest_cuda_yaml_22`. It's ready once it passed the farm.

**Comment by esc** [comment]
`numba_smoketest_cuda_yaml_22` was fine on this PR.

**Pull request merged**